### PR TITLE
Added `AnyEncodable`

### PR DESCRIPTION
**Pull request by NachoSoto**
For [CF-195] and #695.

A future refactor is making all `HTTPRequest`s `Encodable` instead of manual JSON encoding. However, some of our requests have data that's not known at compile time (like `PostReceiptDataOperation`).
To be able to encode that, we need to be able to represent `[String: Any]` as `Encodable`.

[CF-195]: https://revenuecats.atlassian.net/browse/CF-195?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: Purchases/Misc/AnyEncodable.swift [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            try container.encode(dictionary.mapValues(AnyEncodable.init))
+        case let encodable as Encodable:
+            try encodable.encode(to: encoder)
+
+        default:
+            throw EncodingError.invalidValue(
+                self.value,
+                .init(
+                    codingPath: container.codingPath,
+                    debugDescription: "AnyEncodable value cannot be encoded"
```
Comment (aboedo) [inline Purchases/Misc/AnyEncodable.swift:65]: it'd be cool to print out the type here too

**Review by aboedo — APPROVED** [review]

File: Purchases/Misc/AnyEncodable.swift (hunk 1 above)
Comment (NachoSoto) [reply Purchases/Misc/AnyEncodable.swift:65]: I think it would be added automatically? But maybe not. Let me add a test for this code path anyway.

Comment (NachoSoto) [reply Purchases/Misc/AnyEncodable.swift:65]: Yeah this is the whole error:
```
invalidValue(inf, Swift.EncodingError.Context(codingPath: [_JSONKey(stringValue: "a", intValue: nil)], debugDescription: "Unable to encode Double.infinity directly in JSON. Use JSONEncoder.NonConformingFloatEncodingStrategy.convertToString to specify how the value should be encoded.", underlyingError: nil))
```

File: Purchases/Misc/AnyEncodable.swift [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        case let array as [Any?]:
+            try container.encode(array.map(AnyEncodable.init))
+        case let dictionary as [String: Any?]:
+            try container.encode(dictionary.mapValues(AnyEncodable.init))
+        case let encodable as Encodable:
+            try encodable.encode(to: encoder)
+
+        default:
+            throw EncodingError.invalidValue(
+                self.value,
```
Comment (NachoSoto) [inline Purchases/Misc/AnyEncodable.swift:62]: Because the value is added here.

**Pull request merged**